### PR TITLE
Ladybird: Remove requirement to set SERENITY_SOURCE_DIR from browser and headless-browser binaries

### PR DESCRIPTION
**Pull request by ADKaster**
This requires a bunch of CMake/gn changes and a few code changes.

It's "simpler" now. Assuming a +300 line build system change can meet your definition of simpler.

As a side-effect of this change:

- We no longer install more files than are needed to run the browser
- We no longer install files into ${CMAKE_INSTALL_DATADIR}/res, which was annoying for packagers/distros
    -  They are installed in ${CMAKE_INSTALL_DATADIR}/Lagom instead. If CMAKE_INSTALL_DATADIR is not $PREFIX/share though, we get kind of upset.
- Move helper binaries to libexec/ instead of bin/ on non-macOS. In macOS, we make sure they are always in Ladybird.app/Contents/MacOS. This can be controlled via CMAKE_INSTALL_LIBEXECDIR by packagers/distros.

### CODE REVIEW

**Comment by ADKaster** [comment]
Ah, forgot to open as draft. I still need to test this on macOS, sorry for emailz.

File: Ladybird/Utilities.cpp [hunk 1]
```diff
@@ -36,11 +32,11 @@ void platform_init()
 #ifdef AK_OS_MACOS
         return LexicalPath(app_dir).parent().append("Resources"sv).string();
 #else
-        return LexicalPath(app_dir).parent().append("share"sv).string();
+        return LexicalPath(app_dir).parent().append("share/ladybird"sv).string();
```
Comment (ADKaster) [inline Ladybird/Utilities.cpp:35]: I do wonder if share/lagom would be better? Then we could share it with MacPDF/LinuxPDF etc. This has the effect of putting some files in share/ladybird/ladybird, which feels strange :sweat_smile:

File: Ladybird/Utilities.cpp [hunk 2]
```diff
@@ -12,45 +12,61 @@
...
-        if (source_dir) {
-            return ByteString::formatted("{}/Base", source_dir);
-        }
         auto* home = getenv("XDG_CONFIG_HOME") ?: getenv("HOME");
-        VERIFY(home);
-        auto home_lagom = ByteString::formatted("{}/.lagom", home);
-        if (FileSystem::is_directory(home_lagom))
-            return home_lagom;
-        auto app_dir = application_directory().release_value_but_fixme_should_propagate_errors().to_byte_string();
+        if (home != nullptr) {
```
Comment (ADKaster) [inline Ladybird/Utilities.cpp:37]: I was told by the FreeBSD packager that some configs don't have either XDG_CONFIG_HOME or HOME? wild. Anyway, let's make them not have to patch....

**Comment by ADKaster** [comment]
@clausecker this patch is probably interesting to you :)

File: AK/BufferedStream.h [hunk 3]
```diff
@@ -87,7 +87,8 @@ class BufferedHelper {
         auto const candidate = TRY(find_and_populate_until_any_of(candidates, buffer.size()));
 
         if (stream().is_eof()) {
-            if (buffer.size() < m_buffer.used_space()) {
+            if ((candidate.has_value() && candidate->offset + candidate->size > buffer.size())
```
Comment (timschumi) [inline AK/BufferedStream.h:90]: What's the relation between this and `stream().is_eof()`? Why is it only error-worthy that the delimiter is behind the readable buffer size _after_ the source stream is EOF?

Comment (ADKaster) [reply AK/BufferedStream.h:90]: I have no idea honestly. I'm just assuming that the check is reasonable based on the comment below, and extending the check to the full set of possible return values. @sin-ack had some comments in discord but I didn't get the feeling that we came to a conclusion.

File: Ladybird/Utilities.cpp (hunk 2 above)
Comment (clausecker) [reply Ladybird/Utilities.cpp:48]: `HOME` is usually a given, but you cannot assume that any of the `XDG_...` variables are set.  Hence why `StandardPaths::runtime_directory()` failed on FreeBSD before a fallback was added.

File: AK/BufferedStream.h (hunk 3 above)
Comment (ADKaster) [reply AK/BufferedStream.h:90]: Discussed on discord, we should probably extend this check to the non-eof case as well, but that will need some more testing. Added a test as well.

File: Ladybird/Utilities.cpp (hunk 2 above)
Comment (ADKaster) [reply Ladybird/Utilities.cpp:49]: Gotcha, I was misremembering then. I'll keep this change though, it probably doesn't hurt.

**Review by timschumi — APPROVED** [review]
LGTM, although I'm not 100% comfortable in reviewing the CMake changes.

**Pull request merged**